File: hello-python/src/audio_server/handler/websocket_message.py
"""
WebSocket message data classes for JSON serialization/deserialization.
"""

from dataclasses import dataclass, asdict
from typing import Optional
import json
from enum import Enum


class MessageType(str, Enum):
    """
    WebSocket message types enum.
    All type values are uppercase as per protocol specification.
    """
    START = "START"
    STARTED = "STARTED"
    STOP = "STOP"
    STOPPED = "STOPPED"
    GET = "GET"
    ERROR = "ERROR"
    CONNECTED = "CONNECTED"

    @classmethod
    def from_string(cls, value: Optional[str]) -> Optional['MessageType']:
        """
        Parse string to MessageType enum.
        Case-insensitive comparison for backward compatibility.
        """
        if value is None:
            return None
        try:
            return cls[value.upper()]
        except KeyError:
            return None


@dataclass
class WebSocketMessage:
    """
    WebSocket control message for client-server communication.
    Used for all control messages including START, STOP, GET, and responses.
    """
    type: str
    streamId: Optional[str] = None
    offset: Optional[int] = None
    length: Optional[int] = None
    message: Optional[str] = None

    def to_json(self) -> str:
        """Convert to JSON string, excluding None values."""
        data = {k: v for k, v in asdict(self).items() if v is not None}
        return json.dumps(data)

    def to_dict(self) -> dict:
        """Convert to dictionary, excluding None values."""
        return {k: v for k, v in asdict(self).items() if v is not None}

    @classmethod
    def from_json(cls, json_str: str) -> 'WebSocketMessage':
        """Parse from JSON string."""
        data = json.loads(json_str)
        return cls.from_dict(data)

    @classmethod
    def from_dict(cls, data: dict) -> 'WebSocketMessage':
        """Parse from dictionary."""
        return cls(
            type=data.get("type", ""),
            streamId=data.get("streamId"),
            offset=data.get("offset"),
            length=data.get("length"),
            message=data.get("message")
        )

    def get_message_type(self) -> Optional[MessageType]:
        """
        Get the message type as MessageType enum.
        """
        return MessageType.from_string(self.type)

    @classmethod
    def started(cls, stream_id: str, message: str = "Stream started successfully") -> 'WebSocketMessage':
        """Create a STARTED response message."""
        return cls(type="STARTED", streamId=stream_id, message=message)

    @classmethod
    def stopped(cls, stream_id: str, message: str = "Stream finalized successfully") -> 'WebSocketMessage':
        """Create a STOPPED response message."""
        return cls(type="STOPPED", streamId=stream_id, message=message)

    @classmethod
    def error(cls, message: str) -> 'WebSocketMessage':
        """Create an ERROR response message."""
        return cls(type="ERROR", message=message)

File: hello-typescript/src/server/handler/WebSocketMessage.ts
/**
 * WebSocket message class for JSON serialization/deserialization.
 * Used for all control messages between client and server.
 */

/**
 * WebSocket message types enum.
 * All type values are uppercase as per protocol specification.
 */
export enum MessageType {
  START = "START",
  STARTED = "STARTED",
  STOP = "STOP",
  STOPPED = "STOPPED",
  GET = "GET",
  ERROR = "ERROR",
  CONNECTED = "CONNECTED",
}

/**
 * Parse string to MessageType enum.
 * Case-insensitive comparison for backward compatibility.
 */
export function getMessageType(
  value: string | undefined,
): MessageType | undefined {
  if (!value) return undefined;
  const upperValue = value.toUpperCase();
  return Object.values(MessageType).find((type) => type === upperValue);
}

export interface IWebSocketMessage {
  type: string;
  streamId?: string;
  offset?: number;
  length?: number;
  message?: string;
}

export class WebSocketMessage implements IWebSocketMessage {
  type: string;
  streamId?: string;
  offset?: number;
  length?: number;
  message?: string;

  constructor(
    type: string,
    streamId?: string,
    offset?: number,
    length?: number,
    message?: string,
  ) {
    this.type = type;
    this.streamId = streamId;
    this.offset = offset;
    this.length = length;
    this.message = message;
  }

  /**
   * Convert to JSON string, excluding undefined values.
   */
  toJSON(): string {
    const obj: IWebSocketMessage = { type: this.type };
    if (this.streamId !== undefined) obj.streamId = this.streamId;
    if (this.offset !== undefined) obj.offset = this.offset;
    if (this.length !== undefined) obj.length = this.length;
    if (this.message !== undefined) obj.message = this.message;
    return JSON.stringify(obj);
  }

  /**
   * Parse from JSON string.
   */
  static fromJSON(jsonStr: string): WebSocketMessage {
    const data = JSON.parse(jsonStr);
    return WebSocketMessage.fromObject(data);
  }

  /**
   * Parse from object.
   */
  static fromObject(data: IWebSocketMessage): WebSocketMessage {
    return new WebSocketMessage(
      data.type || "",
      data.streamId,
      data.offset,
      data.length,
      data.message,
    );
  }

  /**
   * Create a STARTED response message.
   */
  static started(
    streamId: string,
    message: string = "Stream started successfully",
  ): WebSocketMessage {
    return new WebSocketMessage(
      "STARTED",
      streamId,
      undefined,
      undefined,
      message,
    );
  }

  /**
   * Create a STOPPED response message.
   */
  static stopped(
    streamId: string,
    message: string = "Stream finalized successfully",
  ): WebSocketMessage {
    return new WebSocketMessage(
      "STOPPED",
      streamId,
      undefined,
      undefined,
      message,
    );
  }

  /**
   * Create an ERROR response message.
   */
  static error(message: string): WebSocketMessage {
    return new WebSocketMessage(
      "ERROR",
      undefined,
      undefined,
      undefined,
      message,
    );
  }
}
